django/contrib/admin/templatetags/admin_list.py

```diff
@@ -367,8 +367,7 @@ def link(filters):
                 'choices': [{'title': capfirst(formats.date_format(day, 'MONTH_DAY_FORMAT'))}]
             }
         elif year_lookup and month_lookup:
-            days = cl.queryset.filter(**{year_field: year_lookup, month_field: month_lookup})
-            days = getattr(days, 'dates')(field_name, 'day')
+            days = getattr(cl.queryset, 'dates')(field_name, 'day')
```

Comment:
Shouldn't we keep filtering the queryset in both cases but use the `__gte`/`__lt` approach instead? Were these filters superfluous the whole time?

Reply:
`date_heirarchy` oprerate on the filtered queryset so at the point the query is already filtered. TBH I wasn't sure why it was repplied in the first place.

Am I missing something?

Reply:
> Am I missing something?

Probably not but we'd need to assert this works correctly and I'm not sure the current test coverage accounts for that. We maybe have some assertions against it in `admin_views` tests?

Reply:
In all cases, if the `django/contrib/admin/templatetags/admin_list.py` changes are not required to fix this PR then they should be moved to another commit which can live in this PR as well.

e.g. `Refs #28933 -- Removed unnecessary filtering in date_hierarchy template tag.`

Reply:
OK, seperated the commits. 

The first commit also adds tests to `date_heirarchy`.